Enhancement: Added file size and subdirectory support to FileManager directory listing feature

Isaac-G5900 opened the pull request:
| | |
|:---|:---|
|**_Related Issue(s)_**| #1918  |
|**_Has Unit Tests (y/n)_**| Y |
|**_Documentation Included (y/n)_**| Y (in code comments) |
|**_Generative AI was used in this contribution (y/n)_**| Y |

---
## Change Description

This PR adds a directory listing feature to the FileManager component. The implementation includes:
- New command to list directory contents
- Support for displaying file sizes
- Identification of subdirectories vs. regular files
- Proper error handling and reporting
- Unit tests for the new functionality

## Rationale

This enhancement improves visibility into the file system structure. The directory listing feature with size and subdirectory information allows for better file system management and debugging capabilities during operations.

## Testing/Review Recommendations

Please review the following:
1. ListDirectory_cmdHandler implementation in FileManager.cpp for correct behavior
2. Error handling for edge cases (empty directories, permission issues)
3. Unit tests covering successful listings and failure cases
4. Unit tests for subdirectory identification

All unit tests pass successfully. The implementation was tested with various directory structures to ensure proper handling of files and subdirectories.

## Future Work

Potential future enhancements could include:
- Recursive directory listing capability
- Filtering options by file type or name pattern
- More detailed file metadata in listings

## AI Usage (see [policy](https://github.com/nasa/fprime/blob/devel/AI_POLICY.md))

I used AI to assist in comprehending the fprime codebase, understanding the unit test framework, and composing the unit tests and feature implementation.

Isaac-G5900 committed:
Enhancement: Added file size and subdirectory support to FileManager directory listing feature

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,14 +156,128 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+    Fw::String filename;
+    dir.rewind(); //start from beginning
+    
+    while (true) {
```
Flight standards expect this to be a bounded loop.

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,14 +156,128 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+            this->log_ACTIVITY_HI_DirectoryListing(logStringDirName, logStringFileName, index, displaySize);
+        } else if (pathType == Os::FileSystem::DIRECTORY) {
+            // Subdirectory: count files and emit subdirectory event
+            U32 subdirFileCount = this->countFilesInDirectory(fullPath.toChar());
```
Do we need this functionality?

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,14 +156,128 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+    Fw::String filename;
+    dir.rewind();
+    
+    while (true) {
```
Assuming we keep this, we need to bound this loop too.

LeStarch commented on Svc/FileManager/FileManager.hpp:
```diff
@@ -14,6 +14,7 @@
 #define Svc_FileManager_HPP
 
 #include "Os/FileSystem.hpp"
+#include "Os/Directory.hpp"
```
I don't see this being used in the .hpp, probably should move it to the .cpp file.

LeStarch reviewed (changes requested):
A few minor changes are needed to make this flight standard.  However, my biggest concern is that a large directory will hammer the event subsystem.  At best you drop listing events.  At worst, you'll crash something.

We probably need to implement this via a rategroup and emit one event per rate tick to ensure we don't overload down stream components.  @timcanham  what are your thoughts here?

Isaac-G5900 committed:
Refactor to directory listing feature using rategroup 2 component

Isaac-G5900 commented:
> A few minor changes are needed to make this flight standard. However, my biggest concern is that a large directory will hammer the event subsystem. At best you drop listing events. At worst, you'll crash something.
> 
> We probably need to implement this via a rategroup and emit one event per rate tick to ensure we don't overload down stream components. @timcanham what are your thoughts here?

I've pushed an updated implementation addressing your concerns about the event flooding. The current refactor now uses the rategroup component on rategroup 2 per @LeStarch's suggestion. Rather than the unbounded loops the command now is managed with a state machine. 

Key Changes:
-Rate Group Integration: Connected to Rate Group 2 via schedIn port
-Bounded Execution: Processes exactly one directory entry per rate tick (every 2 seconds)
-State Machine: Clean IDLE/LISTING_IN_PROGRESS states with proper cleanup
-No Event Flooding: Events are naturally spaced by the rate group timing

Along with these changes are unit tests that validate expected behavior. 

Regarding the subdir functionality, I thought it may be more informative handling subdir cases distinctly but am open to simplifying/omitting it per @timcanham's thoughts on the overall approach. 

Currently, the subdirlisting test cases anticipate file counts and have left in placeholder logic until further input.

LeStarch commented:
The last comment is that we should probably make the count of events downlinked per rate group call configurable.  That is, allow multiple files per call.

Config files go in `default/config`  make one for FileManager, and add a loop to downlink using that const.

Are you up to make that change?  Also we need to fix spelling:

```
binaryfile
emptydir
LISTDIRECTORY
```
Add these to `.github/actions/spelling/expect.txt`

Very close, and awesome work!

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+
+@ Starting directory listing
+event ListDirectoryStarted(
+                            dirName: string size 256 @< The name of the directory
```
Instead of 256, use `FileNameStringSize` it is a FPP constant for file strings

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+
+@ Directory listing completed successfully  
+event ListDirectorySucceeded(
+                              dirName: string size 256 @< The name of the directory
```
Same here `FileNameStringSize`

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+
+@ Failed to list directory
+event ListDirectoryError(
+                          dirName: string size 256 @< The name of the directory
```
And here `FileNameStringSize`

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+
+@ Directory listing file entry
+event DirectoryListing(
+                        dirName: string size 256 @< The name of the directory
```
`FileNameStringSize`

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+@ Directory listing file entry
+event DirectoryListing(
+                        dirName: string size 256 @< The name of the directory
+                        fileName: string size 256 @< The name of the file
```
`FileNameStringSize`

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+
+@ Directory listing subdirectory entry
+event DirectoryListingSubdir(
+                        dirName: string size 256 @< The name of the directory
```
`FileNameStringSize`

LeStarch commented on Svc/FileManager/Events.fppi:
```diff
@@ -179,3 +179,51 @@ event FileSizeStarted(
+@ Directory listing subdirectory entry
+event DirectoryListingSubdir(
+                        dirName: string size 256 @< The name of the directory
+                        subdirName: string size 256 @< The name of the subdirectory
```
`FileNameStringSize`

LeStarch commented on Svc/FileManager/test/ut/FileManagerTester.cpp:
```diff
@@ -363,6 +363,169 @@ void FileManagerTester ::fileSizeFail() {
+    this->system("touch test_dir/file2.txt");
+    this->system("touch test_dir/file3.dat");
+#else
+    FAIL();  // Commands not implemented for this OS
```
My preference is to use `SKIP()` here, because an unimplemented unit test should not result in a failure.

LeStarch commented on Svc/FileManager/test/ut/FileManagerTester.cpp:
```diff
@@ -363,6 +363,169 @@ void FileManagerTester ::fileSizeFail() {
+    ASSERT_GT(this->eventHistory_DirectoryListingSubdir->size(), 0U);
+    
+    // We need to skip the assertEvents method to avoid the event count check
+    return;
```
Everything past this is dead code.  This return should be removed, or the the below code should be.

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+    
+    // Check if we're already listing a directory
+    if (m_listState == LISTING_IN_PROGRESS) {
+        this->log_WARNING_HI_ListDirectoryError(logStringDirName, static_cast<U32>(Os::Directory::OTHER_ERROR));
```
Minor: can you pass `dirName` straight in without making a `logStringDirName` variant?  I think that should work now.  Note: this is a new feature.

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+
+    // Initialize state machine for asynchronous processing
+    m_listState = LISTING_IN_PROGRESS;
+    m_currentDirName = dirName.toChar();
```
Shouldn't need the `toChar` here do we?

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+                // Regular file: get size and emit file event
+                FwSizeType fileSize;
+                Os::FileSystem::Status sizeStatus = Os::FileSystem::getFileSize(fullPath.toChar(), fileSize);
+                U64 displaySize = (sizeStatus == Os::FileSystem::OP_OK) ? static_cast<U64>(fileSize) : 0;
```
`FwSizeType` can now be used in events!  We should switch to that and remove the display size.

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+        Fw::String filename;
+        Os::Directory::Status status = m_currentDir.read(filename);
+        
+        if (status == Os::Directory::NO_MORE_FILES) {
```
This is where to insert the for loop from 0 to configured Events per RG tick

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+    m_currentIndex = 0;
+    m_totalEntries = 0;
+    
+    // Directory listing will be processed asynchronously by Rate Group 2 (0.5Hz).
```
RGs are configurable per project.  Remove the `2 and (0.5Hz)` or note it as an example.

LeStarch commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -156,10 +164,106 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+                this->log_ACTIVITY_HI_DirectoryListing(logStringDirName, logStringFileName, m_currentIndex, 0);
+            }
+            
+            m_currentIndex++;
```
Do we need the index of the file listing?  Seems like it is extra unneeded information.

LeStarch commented on Svc/FileManager/FileManager.hpp:
```diff
@@ -132,6 +173,41 @@ class FileManager final : public FileManagerComponentBase {
+    Fw::String m_currentDirName;
+
+    //! Current index in directory listing (for event reporting)
+    U32 m_currentIndex;
```
Again, I'd switch to FwSizeType as it is now supported in the Events!

LeStarch reviewed (changes requested):
This is looking very good!  Thanks for the contribution!

To close out:
1. Add spelling fixes to `expect.txt`
2. Add a configuration field for files-per-schedin call and rategroup
3. Fix minor comments and types

If you need assistance with any of this, please let us know!

Isaac-G5900 committed:
Spelling fixes to expect.txt, config feilds for schedin call, fixes on comments/types

Isaac-G5900 committed:
Fix for spellcheck

Isaac-G5900 commented:
> This is looking very good! Thanks for the contribution!
> 
> To close out:
> 
> 1. Add spelling fixes to `expect.txt`
> 2. Add a configuration field for files-per-schedin call and rategroup
> 3. Fix minor comments and types
> 
> If you need assistance with any of this, please let us know!

Pushed changes once more, Let me know if there was anything I missed from list of last-tweaks suggested. @LeStarch

LeStarch committed:
Replace fixed size with FileNameStringSize for commands

LeStarch commented:
Overall it looks good.  Letting CI run.  I pushed a fix to make commands use FilenameSize constant.

LeStarch committed:
Fix minor nits

LeStarch committed:
Formatting

LeStarch commented:
@Isaac-G5900 I made a few edits to close this PR off.  These were minor nits that I did not want to pester you with.  So you can understand what I did:

1. I removed your test methods and replaced them with the F Prime patterns of: friend class, and invok_to + doDispatch.
2. I updated event and command sizes to use that string variable
3. I updated the CMake to use the `UT_AUTO_HELPERS` to attach the `schedIn` port automatically.

Ran the UTs and saw fairly good coverage.  CI is breaking, so I have a few more fixes to come.  We'll get it merged soon!

LeStarch committed:
Remove fileManager instance, use FileHandling.fileManager

LeStarch commented:
The last change I needed to make was to revert Ref to use `FileHandling.fileManager` as we are shipping this in a reusable subtopology.

Overall:
1. I **really** appreciate your patience going through this.  You've responded to all our requests for minor/major design and implementation changes!
2. I appreciate you tackling a back-logged issue.
3. This work is in great shape.  🤞 that CI passes.

LeStarch reviewed (dismissed):
LGTM, waiting on CI to pass!

thomas-bc commented on default/config/FileManagerConfig.hpp:
```diff
@@ -0,0 +1,16 @@
+#include <config/FpConfig.hpp>
+
+namespace Default {
+    namespace Config {
```
`Default::Config` namespace is not a pattern we seem to use. I'd recommend sticking to the same style that other configs use, which seems to be to put everything in the top level namespace.

thomas-bc commented on Svc/FileManager/FileManager.hpp:
```diff
@@ -91,12 +93,29 @@ class FileManager final : public FileManagerComponentBase {
     );
 
+    //! Handler implementation for schedIn
+    //! This handler is called by Rate Group 2 (0.5Hz) to process
```
nit: which rate group is connected (and its rate) would be project-specific

thomas-bc commented on Svc/FileManager/Events.fppi:
```diff
@@ -174,8 +174,54 @@ event FileSizeError(
+                            ) \
+  severity activity high \
+  id 0x16 \
+  format "Listed {} files in directory {}"
```
This format string read backwards

<img width="381" height="62" alt="Image" src="https://github.com/user-attachments/assets/b46f3b23-6442-485d-84dc-cedcf7fbde61" />

thomas-bc commented on Svc/FileManager/FileManager.cpp:
```diff
@@ -149,17 +157,109 @@ void FileManager ::FileSize_cmdHandler(const FwOpcodeType opCode, const U32 cmdS
+                        (sizeStatus == Os::FileSystem::OP_OK) ? fileSize : static_cast<FwSizeType>(0));
+                } else if (pathType == Os::FileSystem::DIRECTORY) {
+                    // Subdirectory: emit subdirectory event with file count 0 (simplified)
+                    this->log_ACTIVITY_HI_DirectoryListingSubdir(m_currentDirName, filename, 0);
```
Why include a file count argument to the event `DirectoryListingSubdir` if it's always going to report 0 ? I'd rather remove it from the event and not report at all ?

Isaac-G5900 replied:
Yes, I agree. Just an artifact of several changes. Initial thought was to at least treat the subdir events with a file count if only interested in listing a given directory. 

If something like that is desired, I’m open to implementing that. Whatever you and the team believe is best.

thomas-bc reviewed (commented):
A few nits and one question about the event argument

LeStarch commented:
@Isaac-G5900 do you want to close-out @thomas-bc requests or shall I?

Isaac-G5900 commented:
> @Isaac-G5900 do you want to close-out @thomas-bc requests or shall I?

I think you should for the sake of ensuring it's done as desired with less iteration and putting the PR to rest. I will glean the benefit of an example of best practices and incorporate them in future contributions. I greatly appreciate the learning opportunity and patience. Cheers.

LeStarch committed:
Fix review feedback

LeStarch commented:
@Isaac-G5900 done.  I never want to step on other's work, but also understand that sometimes the polishing of a contribution can get iteration-heavy.

@thomas-bc I liked the configuration style (namespaced) but I swapped the tokens.  The other issues should be fixed.
